core/src/storage2/lazy/lazy_array.rs

```diff
@@ -0,0 +1,444 @@
+            }
+        }
+    }
+
+    /// Loads the entry at the given index.
+    ///
+    /// Tries to load the entry from cache and falls back to lazily load the
+    /// entry from the contract storage.
+    ///
+    /// # Panics
+    ///
+    /// - If the lazy array is in a state that forbids lazy loading.
+    /// - If the given index is out of bounds.
+    fn load_through_cache_mut(&mut self, index: Index) -> &mut Entry<T> {
+        // SAFETY:
+        // Returning a `&mut Entry<T>` from within a `&mut self` function
+        // won't allow creating aliasing between exclusive references.
+        unsafe { &mut *self.load_through_cache(index).as_ptr() }
+    }
+
+    /// Returns a shared reference to the element at the given index if any.
+    ///
+    /// # Note
+    ///
+    /// This operation eventually loads from contract storage.
+    ///
+    /// # Panics
+    ///
+    /// If the given index is out of bounds.
+    pub fn get(&self, at: Index) -> Option<&T> {
+        unsafe { &*self.load_through_cache(at).as_ptr() }
+            .value()
+            .into()
+    }
+
+    /// Returns an exclusive reference to the element at the given index if any.
+    ///
+    /// # Note
+    ///
+    /// This operation eventually loads from contract storage.
+    ///
+    /// # Panics
+    ///
+    /// If the given index is out of bounds.
+    pub fn get_mut(&mut self, at: Index) -> Option<&mut T> {
+        self.load_through_cache_mut(at).value_mut().into()
+    }
+
+    /// Removes the element at the given index and returns it if any.
+    ///
+    /// # Note
+    ///
+    /// This operation eventually loads from contract storage.
+    ///
+    /// # Panics
+    ///
+    /// If the given index is out of bounds.
+    pub fn take(&mut self, at: Index) -> Option<T> {
+        self.load_through_cache_mut(at).take_value()
+    }
+
+    /// Puts the new value into the indexed slot and returns the old value if any.
+    ///
+    /// # Note
+    ///
+    /// - This operation eventually loads from contract storage.
+    /// - Prefer [`LazyArray::put`] if you are not interested in the old value.
+    /// - Use [`LazyArray::put_get`]`(None)` to remove an element.
+    ///
+    /// # Panics
+    ///
+    /// If the given index is out of bounds.
+    pub fn put_get(&mut self, at: Index, new_value: Option<T>) -> Option<T> {
+        self.load_through_cache_mut(at).put(new_value)
+    }
+
+    /// Swaps the values at indices x and y.
+    ///
+    /// # Note
+    ///
+    /// This operation eventually loads from contract storage.
+    ///
+    /// # Panics
+    ///
+    /// If any of the given indices is out of bounds.
+    pub fn swap(&mut self, a: Index, b: Index) {
+        if a == b {
+            // Bail out early if both indices are the same.
+            return
+        }
+        assert!(a < Self::capacity(), "a is out of bounds");
+        assert!(b < Self::capacity(), "b is out of bounds");
+        let (loaded_a, loaded_b) =
```

Comment:
`cargo miri test` reports undefined behaviour at this line of code.
I have already checked it and am pretty confident about it since it is very similar to the `swap` implementation of `storage2::Vec` and that one doesn't trigger undefined behavior according to `miri`.
It would be very nice if reviewers could be extra careful about this code to spot potential pitfalls.

The `miri` error is the following:

```
error: Undefined Behavior: trying to reborrow for Unique, but parent tag <2220142> does not have an appropriate item in the borrow stack
   --> core/src/storage2/lazy/lazy_array.rs:426:14
    |
426 |         let (loaded_a, loaded_b) =
    |              ^^^^^^^^ trying to reborrow for Unique, but parent tag <2220142> does not have an appropriate item in the borrow stack
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
    = note: inside `storage2::lazy::lazy_array::LazyArray::<u8, typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UTerm, typenum::bit::B1>, typenum::bit::B0>, typenum::bit::B0>>::swap` at core/src/storage2/lazy/lazy_array.rs:426:14
note: inside `storage2::collections::smallvec::SmallVec::<u8, typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UTerm, typenum::bit::B1>, typenum::bit::B0>, typenum::bit::B0>>::swap` at core/src/storage2/collections/smallvec/mod.rs:247:9
   --> core/src/storage2/collections/smallvec/mod.rs:247:9
    |
247 |         self.elems.swap(a, b)
    |         ^^^^^^^^^^^^^^^^^^^^^
note: inside `storage2::collections::smallvec::tests::swap_works` at core/src/storage2/collections/smallvec/tests.rs:273:5
   --> core/src/storage2/collections/smallvec/tests.rs:273:5
    |
273 |     vec.swap(0, 1);
    |     ^^^^^^^^^^^^^^
note: inside closure at core/src/storage2/collections/smallvec/tests.rs:262:1
   --> core/src/storage2/collections/smallvec/tests.rs:262:1
    |
262 | / fn swap_works() {
263 | |     let elems = [b'A', b'B', b'C', b'D'];
264 | |     let mut vec = vec_from_slice(&elems);
265 | |
...   |
280 | |     assert_eq_slice(&vec, &[b'C', b'A', b'D', b'B']);
281 | | }
    | |_^
    = note: inside `<[closure@core/src/storage2/collections/smallvec/tests.rs:262:1: 281:2] as std::ops::FnOnce<()>>::call_once - shim` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:232:5
    = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:232:5
    = note: inside `test::__rust_begin_short_backtrace::<fn()>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:517:5
    = note: inside closure at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:508:30
    = note: inside `<[closure@DefId(38:631 ~ test[140d]::run_test[0]::{{closure}}[2]) 0:fn()] as std::ops::FnOnce<()>>::call_once - shim(vtable)` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:232:5
    = note: inside `<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send> as std::ops::FnOnce<()>>::call_once` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/liballoc/boxed.rs:1008:9
    = note: inside `<std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>> as std::ops::FnOnce<()>>::call_once` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:318:9
    = note: inside `std::panicking::r#try::do_call::<std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>>, ()>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:331:40
    = note: inside `std::panicking::r#try::<(), std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>>>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:274:15
    = note: inside `std::panic::catch_unwind::<std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>>, ()>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:394:14
    = note: inside `test::run_test_in_process` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:541:18
    = note: inside closure at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:450:39
    = note: inside `test::run_test::run_test_inner` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:475:13
    = note: inside `test::run_test` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:505:28
    = note: inside `test::run_tests::<[closure@DefId(38:230 ~ test[140d]::console[0]::run_tests_console[0]::{{closure}}[2]) 0:&mut test::console::ConsoleTestState, 1:&mut std::boxed::Box<dyn test::formatters::OutputFormatter>]>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:284:13
    = note: inside `test::run_tests_console` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/console.rs:280:5
    = note: inside `test::test_main` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:120:15
    = note: inside `test::test_main_static` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libtest/lib.rs:139:5
    = note: inside `main`
    = note: inside closure at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:34
    = note: inside closure at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:73
    = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<[closure@DefId(1:6025 ~ std[1995]::rt[0]::lang_start_internal[0]::{{closure}}[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/sys_common/backtrace.rs:130:5
    = note: inside closure at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:13
    = note: inside `std::panicking::r#try::do_call::<[closure@DefId(1:6024 ~ std[1995]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:331:40
    = note: inside `std::panicking::r#try::<i32, [closure@DefId(1:6024 ~ std[1995]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe]>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:274:15
    = note: inside `std::panic::catch_unwind::<[closure@DefId(1:6024 ~ std[1995]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:394:14
    = note: inside `std::rt::lang_start_internal` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:51:25
    = note: inside `std::rt::lang_start::<()>` at /home/robrob/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:5
    = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```